src/throttler/throttler.ts

```diff
@@ -254,4 +230,79 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(taskData: TaskData<T, R>, cursorIndex: number): Promise<void> {
+    let rejectFunc: (err: TaskError) => void;
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      rejectFunc = reject;
+    });
+    taskData.timeoutId = setTimeout(() => {
+      taskData.isTimedOut = true;
+      rejectFunc(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+    }, timeoutMillis);
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(taskData: TaskData<T, R>, cursorIndex: number): Promise<any> {
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
+      const dt = Date.now() - t0;
+      if (dt < this.min) {
+        this.min = dt;
+      }
+      if (dt > this.max) {
+        this.max = dt;
+      }
+      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
+
+      this.success++;
+      this.complete++;
+      this.active--;
+      return result;
+    } catch (error) {
+      if (taskData.retryCount === this.retries) {
+        throw new RetriesExhaustedError(this.taskName(cursorIndex), this.retries, error);
+      }
+      if (taskData.isTimedOut) {
+        throw error;
+      }
+      this.retried++;
+      taskData.retryCount++;
+      await backoff(taskData.retryCount, this.backoff);
+      logger.debug(`[${this.name}] Retrying task`, this.taskName(cursorIndex));
+      return this.executeTask(taskData, cursorIndex);
+    }
+  }
+
+  private onTaskFulfilled(result: any, taskData: TaskData<T, R>, cursorIndex: number): void {
+    if (taskData.wait) {
+      taskData.wait.resolve(result);
+    }
+    this.cleanupTask(taskData, cursorIndex);
+    this.process();
+  }
+
+  private onTaskFailed(error: TaskError, taskData: TaskData<T, R>, cursorIndex: number): void {
+    this.errored++;
+    this.complete++;
+    this.active--;
+
+    logger.debug(error);
+
+    if (taskData.wait) {
+      taskData.wait.reject(error);
+    }
+    this.cleanupTask(taskData, cursorIndex);
+    this.finish(error);
```

Comment:
I think this should be `this.finishIfIdle()`, since it's used for marking all tasks in the throttler as finished, rather than individual tasks.

Reply:
I don't think `finishIfIdle()` does that. It simply calls `finish()` if the queue is closed and there's no waiting or running task.
`finish()` also doesn't mark tasks as finished. It tries to notify all the promises returned from `wait()` calls that the `Throttler` already completed all tasks (resolve), or report the error as soon as a task fails (reject). 
Here we want to `finish(error)` since we want to reject all the current `wait()` promise with the error

Reply:
Ah yep, you're right. One task failing causing the whole throttler to immediately fail matches the previous behaviour.